File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int mk = Integer.parseInt(st.nextToken()) + Integer.parseInt(st.nextToken());

        SegmentTree segTree = new SegmentTree(n);

        for (int i = 0; i < n; i++) {
            segTree.update(i, Long.parseLong(br.readLine()));
        }

        while (mk-- > 0) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            long b = Long.parseLong(st.nextToken());
            long c = Long.parseLong(st.nextToken());

            if (a == 1) {
                segTree.update((int) b - 1, c);
                continue;
            }

            bw.append(String.format("%d\n", segTree.query((int) b - 1, (int) c - 1)));
        }

        bw.flush();
        bw.close();
    }
}

class SegmentTree {
    long[] tree;
    int cap = 1;

    SegmentTree(int n) {
        while (cap < n) cap <<= 1;

        tree = new long[cap << 1];
    }

    void update(int i, long v) {
        int node = cap + i;
        long w = v - tree[node];

        while (node > 0) {
            tree[node] += w;
            node >>>= 1;
        }
    }

    long query(int b, int c) {
        long res = 0;
        int s = cap + b;
        int e = cap + c;

        while (s < e) {
            if ((s & 1) == 1) res += tree[s++];
            if ((e & 1) == 0) res += tree[e--];

            s >>>= 1;
            e >>>= 1;
        }

        if (s == e) res += tree[s];

        return res;
    }
}

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.kt
import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val bw = System.out.bufferedWriter()
    val (n, mk) = StringTokenizer(readLine()).run {
        val token = { nextToken().toInt() }
        token() to token() + token()
    }
    val st = SegmentTree(n)

    repeat(n) { st.update(it, readLine().toLong()) }

    repeat(mk) {
        val line = StringTokenizer(readLine())
        val token = { line.nextToken().toLong() }

        when (token()) {
            1L -> st.update(token().toInt() - 1, token())
            2L -> bw.append("${st.query(token().toInt() - 1, token().toInt() - 1)}\n")
        }
    }
    bw.flush()
    bw.close()
    close()
}

class SegmentTree(n: Int) {
    val tree: LongArray
    var cap = 1

    init {
        while (cap < n) cap *= 2

        tree = LongArray(cap shl 1)
    }

    fun update(i: Int, v: Long) {
        var node = cap + i
        val weight = v - tree[node]

        while (node > 0) {
            tree[node] += weight
            node = node ushr 1
        }
    }

    fun query(qs: Int, qe: Int): Long {
        var ret = 0L
        var s = qs + cap
        var e = qe + cap

        while (s < e) {
            if (s and 1 == 1) ret += tree[s++]

            if (e and 1 == 0) ret += tree[e--]

            s = s ushr 1
            e = e ushr 1
        }
        if (s == e) ret += tree[s]

        return ret
    }
}